File: 547-number-of-provinces/number-of-provinces.cpp
class Solution {
public:
    // void dfs(vector<vector<int>>&adj,vector<int>&visited,int &src){
    //     visited[src]=1;
    //     for(int i:adj[src]){
    //         if(!visited[i])
    //             dfs(adj,visited,i);
    //     }
    // }
    void bfs(vector<vector<int>>&adj,vector<int>&visited,int &src){

        queue<int>q;

        q.push(src);
        while(!q.empty()){

            int curr_n=q.front();
            visited[curr_n]=1;
            q.pop();

            for(int i:adj[curr_n]){
                if(!visited[i]){
                    visited[i]=1;
                    q.push(i);
                }
            }
        }
    }
    int findCircleNum(vector<vector<int>>& isConnected) {
        
        int v= isConnected.size();
        vector<vector<int>>adj(v,vector<int>(v));
        for(int i=0;i<v;i++){
            for(int j=0;j<v;j++){

                if(isConnected[i][j]){
                    int u=i;
                    int V=j;
                    adj[u].push_back(V);
                    adj[V].push_back(u);
                }
            }
        }

        vector<int>visited(v,0);
        int result=0;
        for(int i=0;i<v;i++){
            if(!visited[i]){
                bfs(adj,visited,i);
                result++;
            }
        }
        return result;
    }
};

File: 547-number-of-provinces/number-of-provinces.java
class Solution {
    public int findCircleNum(int[][] isConnected) {
        int n= isConnected.length;
        int m=isConnected[0].length;
        ArrayList<ArrayList<Integer>>adj = new ArrayList<>();

        for(int i=0;i<n;i++){
            adj.add(new ArrayList());
        }

        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(i==j) continue;
                if(isConnected[i][j]==1)
                    adj.get(i).add(j);
            }
        }

        boolean vis[] = new boolean[n+1];
        int result=0;
        for(int i=0;i<n;i++){
            if(!vis[i]){
                result++;
                dfs(i,vis,adj);
            }
        }
        return result;
    }

    public static void dfs(int src,boolean vis[],ArrayList<ArrayList<Integer>>adj){

        vis[src]=true;

        for(int i:adj.get(src)){
            if(!vis[i]){
                dfs(i,vis,adj);
            }
        }
    }
}